consensus/src/simplex/types.rs

```diff
@@ -0,0 +1,1450 @@
+    }
+}
+
+impl<S: Array, D: Digest> Read for Notarize<S, D> {
+    fn read_cfg(reader: &mut impl Buf, _: &()) -> Result<Self, Error> {
+        let proposal = Proposal::<D>::read_cfg(reader, &())?;
+        let signature = Signature::<S>::read_cfg(reader, &())?;
+        Ok(Self {
+            proposal,
+            signature,
+        })
+    }
+}
+
+impl<S: Array, D: Digest> FixedSize for Notarize<S, D> {
+    const SIZE: usize = Proposal::<D>::SIZE + Signature::<S>::SIZE;
+}
+
+impl<S: Array, D: Digest> Viewable for Notarize<S, D> {
+    fn view(&self) -> View {
+        self.proposal.view()
+    }
+}
+
+impl<S: Array, D: Digest> Attributable for Notarize<S, D> {
+    fn signer(&self) -> u32 {
+        self.signature.signer()
+    }
+}
+
+/// Notarization represents an aggregated set of notarizes that meets the quorum threshold.
+/// It includes the proposal and the set of signatures from validators.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Notarization<S: Array, D: Digest> {
+    /// The proposal that has been notarized
+    pub proposal: Proposal<D>,
+    /// The set of signatures from validators (must meet quorum threshold)
+    pub signatures: Vec<Signature<S>>,
+}
+
+impl<S: Array, D: Digest> Notarization<S, D> {
+    /// Creates a new notarization with the given proposal and set of signatures.
+    /// The signatures are sorted by the public key index for deterministic ordering.
+    pub fn new(proposal: Proposal<D>, mut signatures: Vec<Signature<S>>) -> Self {
+        signatures.sort_by_key(|s| s.public_key);
+        Self {
+            proposal,
+            signatures,
+        }
+    }
+
+    /// Verifies all signatures in this notarization using the provided verifier.
+    ///
+    /// This ensures that:
+    /// 1. There are at least threshold valid signatures
+    /// 2. No duplicate signers
+    /// 3. All signatures are valid
+    /// 4. All signers are in the validator set
+    // TODO(#755): Use `commonware-cryptography::Specification`
+    pub fn verify<P: Array, V: Verifier<PublicKey = P, Signature = S>>(
+        &self,
+        namespace: &[u8],
+        participants: &[P],
+    ) -> bool {
+        // Get allowed signers
+        let (threshold, count) = threshold(participants);
+        if self.signatures.len() < threshold as usize {
+            return false;
+        }
+        if self.signatures.len() > count as usize {
+            return false;
+        }
+
+        // Verify signatures
+        let notarize_namespace = notarize_namespace(namespace);
+        let mut last_seen = None;
+        let message = self.proposal.encode();
+        for signature in &self.signatures {
+            // Ensure this isn't a duplicate (and the signatures are sorted)
+            if let Some(last_seen) = last_seen {
+                if last_seen >= signature.public_key {
+                    return false;
+                }
+            }
+            last_seen = Some(signature.public_key);
```

Comment:
Consider using `HashMap<P, S>` (serialization and deserialization ensures keys are unique and sorted), or moving the logic to `read_cfg`

Reply:
Want to avoid the map allocation, so I'll keep ~as is but move to read_cfg 👍